feat: prevent Z-fighting/flicker in OpenSCAD preview

spetzreborn opened the pull request:
This adds the constant EPSILON to prevent Z-fighting / Z-flicker. Could be conceded "code smell" but is well known OpenSCAD approach:

* https://en.wikibooks.org/wiki/OpenSCAD_User_Manual/FAQ#What_are_those_strange_flickering_artifacts_in_the_preview?
* https://en.wikibooks.org/wiki/OpenSCAD_User_Manual/CSG_Modelling (search for eps)

    This makes it easier for the eyes when working in OpenSCAD and preview.

Commit:
feat: prevent Z-fighting/flicker in OpenSCAD preview

This adds the constant EPSILON to prevent Z-fighting / Z-flicker. Could
be conceded "code smell" but is well known OpenSCAD approach:

* https://en.wikibooks.org/wiki/OpenSCAD_User_Manual/FAQ#What_are_those_strange_flickering_artifacts_in_the_preview?
* https://en.wikibooks.org/wiki/OpenSCAD_User_Manual/CSG_Modelling
(search for eps)

    This makes it easier for the eyes when working in OpenSCAD and preview.

kellervater commented on models/core/makerworld/support.scad:
```diff
@@ -375,7 +376,7 @@ module tension_hole_half(){
   lockpin_tension_hole_width_inner = PRINTING_LAYER_WIDTH * 4;
   lockpin_tension_hole_height = BASE_UNIT / 2;
-  lockpin_tension_hole_inner_dimension = [lockpin_tension_hole_width_inner, lockpin_height];
+  lockpin_tension_hole_inner_dimension = [lockpin_tension_hole_width_inner, lockpin_height + EPSILON ];
```
I wonder if there's a openscad formatter 🤔 as pre-commit hook.

spetzreborn commented:
Yeah, 0.01 is the example from the documentation, it's just me that used a smaller in my template, so should be good to change.

The purpose is only to offset a small part that won't show in the final part.

I have no experience with alter the epsilon for rendering, only $fn. I think that is doable, but I'm not sure if it might create bugs down the line when rendering. I'm guessing but like a thin wall that gets solid in render but not preview.

Maybe doing it as simple as a tenary check? It is common to do for $fn:

$fn = $preview ? 32 : 64;

https://en.wikibooks.org/wiki/OpenSCAD_User_Manual/Other_Language_Features#$fn

Maybe do the same for the value for EPSILON?
(Not at my computer right now so can't test it)

Commit:
fix(core): Difference EPSILON in preview and render.

           Increase EPSILON in preview to limit Z-fighting artifacts, and decrease
           when render.

           If EPSILON is 0 it generates an error in BOSL2 math.scad, so I keep it
           small even if rendering.

spetzreborn committed:
Merge branch 'main' into feat/Z-fight

spetzreborn commented:
Made a commit that difference the epsilon value depending if preview or rendering. I'm not sure that actually is beneficial or just a static value. The epsilon value in preview is now a magnitude bigger, still very small.

I did some tries to setting the epslion to 0, but that generated errors in BOSL2 math.scad so I think that is a nogo.

kellervater commented:
> yes, it is most commonly used for the $fn setting as in my example. The library might benefit from setting a lower $fn in preview and a higher in render. Thinking of moving that to constants.scad or main.scad

imho settings like $fn or EPSILON shouldn't be actively tweaked in library files. They should be agnostic to that.
If we want we can add the Epsilon parameter to the lib modules, but the actual setting and $preview toggling should happen in orchestration files (e.g. makerworld exports or presets).


> It might make gridfinity/base.scad render quicker. Have mostly looked in core and not the models yet so not sure.

Just a fair warning: This one is faulty! The filetting on the edges are not quite right, so the base plate generated from this doesn't 100% adhere to the Gridfinity standard.
tl;dr: still WIP

spetzreborn commented:
> imho settings like $fn or EPSILON shouldn't be actively tweaked in library files. They should be agnostic to that. If we want we can add the Epsilon parameter to the lib modules, but the actual setting and $preview toggling should happen in orchestration files (e.g. makerworld exports or presets).
> 

Yes, that seems correct, and the parts/ files.  So it get set by the correct caller in respective use case. This is on me, but I always forget makerworld. As you might have noticed I'm mostly for local generated ^^

> > It might make gridfinity/base.scad render quicker. Have mostly looked in core and not the models yet so not sure.
> 
> Just a fair warning: This one is faulty! The filetting on the edges are not quite right, so the base plate generated from this doesn't 100% adhere to the Gridfinity standard. tl;dr: still WIP

Yeah, I got that. Was just me trying to familiarise myself with the codebase and look around.


I am no OpenSCAD expert, but the cases I'll seen regarding epsilon and the cousin "use a bigger piece to difference()" has been to over cut and dont care between preview and render. If the object that is differneced is a submodule it does not mater that the "diff object" is to big as the submodule will be used as a solid.
In the other case the epsilon is so small so the slicer and 3dprinter just ignore it and it is floored/roofed to the correct altitude.

 I don't know which way is better. Anyhow I'm happy if the general concept is brought in as it makes the customisation in OpenSCAD a better user experience.

spetzreborn committed:
Merge branch 'main' into feat/Z-fight

kellervater reviewed (changes requested):
I would like to see EPSILON as a named parameter in the modules it is being used and ideally set to 0 per default. If that's not possible due to some bosl2 shenanigans, then it shall be set to the smallest value possible as default.

e.g.
```
module connectorArmOuter(is_foot=false)
---> module connectorArmOuter(is_foot=false, EPSILON=0)  # or something very close to 0
```

This is to prevent a breaking change for the pure purpose of devex while giving everyone the opportunity to set the value higher.
